pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/ReferralSourceService.java

```diff
@@ -0,0 +1,408 @@
+                    && !props.get(DBConstants.REFERRAL_SOURCE_ID).equals(NA_REF_SOURCE))) {
+                return false;
+            }
+            log.info("Updating REFERRAL_SOURCE data {}", msg);
+            props.put(DBConstants.REFERRAL_SOURCE_ID, refSourceId);
+            ParticipantDataDao dataDao = new ParticipantDataDao();
+            dataDao.updateParticipantDataColumn(
+                    new ParticipantData.Builder()
+                            .withParticipantDataId(participantData.getParticipantDataId())
+                            .withDdpParticipantId(ddpParticipantId)
+                            .withDdpInstanceId(participantData.getDdpInstanceId())
+                            .withFieldTypeId(RGP_PARTICIPANT_DATA)
+                            .withData(gson.toJson(props))
+                            .withLastChanged(System.currentTimeMillis())
+                            .withChangedBy(userId).build());
+        } catch (JsonSyntaxException | JsonIOException | ClassCastException e) {
+            throw new DsmInternalError("Invalid data format " + msg, e);
+        }
+        return true;
+    }
+
+    protected static Map<String, String> getDataMap(ParticipantData participantData) throws JsonSyntaxException {
+        Optional<String> data = participantData.getData();
+        if (data.isEmpty() || StringUtils.isEmpty(data.get())) {
+            return Collections.emptyMap();
+        }
+
+        return gson.fromJson(data.get(), Map.class);
+    }
+
+    private List<Activities> getParticipantActivities(String ddpParticipantId, String esIndex) {
+        Optional<ElasticSearchParticipantDto> esParticipant =
+                ElasticSearchUtil.getParticipantESDataByParticipantId(esIndex, ddpParticipantId);
+        if (esParticipant.isEmpty()) {
+            throw new ESMissingParticipantDataException("Participant ES data is null for participant " + ddpParticipantId);
+        }
+        return esParticipant.get().getActivities();
+    }
+
+    /**
+     * Derive REFERRAL_SOURCE ID from participant activities
+     *
+     * @param activities list of DSS participant activities
+     * @return a referral source ID
+     * @throws DsmInternalError If the expected referral source values are not present or the referral source mapping is
+     *                          missing or out of sync
+     */
+    public static String deriveReferralSourceId(List<Activities> activities) {
+        return ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(activities));
+    }
+
+    /**
+     * Get ref sources from activities. Specifically get answers to FIND_OUT enrollment question.
+     *
+     * @return list of strings, empty if question or answers not found
+     */
+    protected static List<String> getReferralSources(List<Activities> activities) {
+        Optional<Activities> enrollmentActivity = activities.stream().filter(activity ->
```

Comment:
just observation and looks like existing code , not related to changes in the PR
DTO Activities.java name is confusing . What its representing is one Activity not multiple activities. Name should have been Activity.java

Reply:
The `Activities` class is pretty widely used.
I can put up a separate PR if you would like the class name changed to `Activity`.

Reply:
Thats fine. we dont need to refactor/rename the class.